File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/MuxTracking.java
package com.brentvatne.exoplayer;

import android.content.Context;

import com.facebook.react.bridge.ReadableMap;
import com.google.ads.interactivemedia.v3.api.AdsLoader;
import com.google.android.exoplayer2.ExoPlayer;
import com.google.android.exoplayer2.Player;
import com.mux.stats.sdk.core.model.CustomerPlayerData;
import com.mux.stats.sdk.core.model.CustomerVideoData;
import com.mux.stats.sdk.core.model.CustomerViewData;
import com.mux.stats.sdk.core.model.CustomerData;
import com.mux.stats.sdk.muxstats.MuxStatsExoPlayer;

import java.util.Arrays;
import java.util.HashMap;

public class MuxTracking {

    private final String CONFIG_ENVIRONMENT_KEY = "environment_key";
    private final String CONFIG_PLAYER_NAME = "player_name";
    private final String CONFIG_VIEW_USER_ID = "user_id";
    private final String CONFIG_VIEW_SUBSCRIPTION_DETAIL = "sub_property";
    private final String CONFIG_VIDEO_TITLE = "video_title";
    private final String CONFIG_VIDEO_CONTENT_TYPE = "video_content_type";
    private final String CONFIG_VIDEO_LANGUAGE_CODE = "video_language_code";
    private final String CONFIG_VIEW_SESSION_ID = "view_session_id";


    private final String[] REQUIRED_CONFIG = {
            CONFIG_ENVIRONMENT_KEY
    };

    private final Context activityContext;
    private MuxStatsExoPlayer tracker;

    MuxTracking(Context context) {
        activityContext = context;
    }

    private String getKeyFrom(ReadableMap map, String key, String fallbackValue) throws Exception {
        String value = map.hasKey(key) ? map.getString(key) : null;
        if (value == null) {
            if (Arrays.asList(REQUIRED_CONFIG).contains(key)) {
                throw new Exception("Mux config key " + key +" is required");
            }
            value = fallbackValue;
        }
        return value;
    }

    private CustomerPlayerData getPlayerData(ReadableMap muxConfig) throws Exception {
        CustomerPlayerData customerPlayerData = new CustomerPlayerData();
        customerPlayerData.setEnvironmentKey(getKeyFrom(muxConfig, CONFIG_ENVIRONMENT_KEY, null));

        String playerName = getKeyFrom(muxConfig, CONFIG_PLAYER_NAME, "vimai-mobile-exoplayer");
        customerPlayerData.setPlayerName(playerName);

        String viewerUserId = getKeyFrom(muxConfig, CONFIG_VIEW_USER_ID, null);
        if (viewerUserId != null) {
            customerPlayerData.setViewerUserId(viewerUserId);
        }

        String viewerSubscriptionDetail = getKeyFrom(muxConfig, CONFIG_VIEW_SUBSCRIPTION_DETAIL, null);
        if (viewerUserId != null) {
            customerPlayerData.setSubPropertyId(viewerSubscriptionDetail);
        }

        return customerPlayerData;
    }

    private CustomerVideoData getVideoData(ReadableMap muxConfig) throws Exception {
        CustomerVideoData customerVideoData = new CustomerVideoData();

        String videoTitle = getKeyFrom(muxConfig, CONFIG_VIDEO_TITLE, null);
        if (videoTitle != null) {
            customerVideoData.setVideoTitle(videoTitle);
        }
        String videoContentType = getKeyFrom(muxConfig, CONFIG_VIDEO_CONTENT_TYPE, null);
        if (videoTitle != null) {
            customerVideoData.setVideoContentType(videoContentType);
        }
        String videoLanguageCode = getKeyFrom(muxConfig, CONFIG_VIDEO_LANGUAGE_CODE, null);
        if (videoTitle != null) {
            customerVideoData.setVideoLanguageCode(videoLanguageCode);
        }
        return customerVideoData;
    }

    private CustomerViewData getViewData(ReadableMap muxConfig) throws Exception {
        CustomerViewData customerViewData = new CustomerViewData();

        String viewSession = getKeyFrom(muxConfig, CONFIG_VIEW_SESSION_ID, null);
        if (viewSession != null) {
            customerViewData.setViewSessionId(viewSession);
        }

        return customerViewData;
    }

    public void track(ExoPlayer player, ReadableMap muxConfig) throws IllegalArgumentException {
        this.release();
        String playerName = muxConfig.hasKey(CONFIG_PLAYER_NAME) ? muxConfig.getString(CONFIG_PLAYER_NAME) : "vimai-mobile-exoplayer";
        try {
            CustomerData data = new CustomerData(
                getPlayerData(muxConfig), getVideoData(muxConfig), getViewData(muxConfig)
            );
            tracker = new MuxStatsExoPlayer(
                    activityContext,
                    player,
                    playerName,
                    data
            );
            System.out.println("==== Mux start tracking video");
        }
        catch (Exception e) {
            tracker = null;
            System.out.println("==== Mux error: " + e.getMessage());
            throw new IllegalArgumentException(e.getMessage());
        }

    }

    public void setStreamType(int type) {
        if (tracker != null) {
            tracker.setStreamType(type);
        }
    }

    public void monitorImaAdsLoader(AdsLoader adsLoader) {
        if (tracker != null && adsLoader != null) {
            tracker.monitorImaAdsLoader(adsLoader);
        }
    }

    public void release() {
        if (tracker != null) {
            tracker.release();
        }
    }
}

File: ios/Video/MuxTracking.m
//
//  MuxTracking.m
//  react-native-video-ima
//
//  Created by Khuong on 4/8/21.
//

#import <Foundation/Foundation.h>
@import MUXSDKStats;
#import "MuxTracking.h"
#import <React/RCTLog.h>

@implementation MuxTracking {
    MUXSDKCustomerPlayerData* customerPlayerData;
    MUXSDKCustomerVideoData* customerVideoData;
    MUXSDKCustomerViewData* customerViewData;
}

static NSString * const CONFIG_ENVIRONMENT_KEY = @"environment_key";
static NSString * const CONFIG_PLAYER_NAME = @"player_name";
static NSString * const CONFIG_VIEW_USER_ID = @"user_id";
static NSString * const CONFIG_VIEW_SUBSCRIPTION_DETAIL = @"sub_property";
static NSString * const CONFIG_VIDEO_TITLE = @"video_title";
static NSString * const CONFIG_VIDEO_CONTENT_TYPE = @"video_content_type";
static NSString * const CONFIG_VIDEO_LANGUAGE_CODE = @"video_language_code";
static NSString * const CONFIG_VIEW_SESSION_ID = @"view_session_id";

static NSString * const REQUIRED_CONFIG[] = {CONFIG_ENVIRONMENT_KEY};

- (NSString*)getKeyFrom:(NSDictionary *)muxConfig key:(NSString*)key fallback:(NSString*)fallback {
    NSUInteger count = sizeof(REQUIRED_CONFIG) / sizeof( REQUIRED_CONFIG[0]);
    NSArray * requiredConfig = [NSArray arrayWithObjects:REQUIRED_CONFIG count:count];
    if ([muxConfig objectForKey:key]) {
        return muxConfig[key];
    } else {
        if ([requiredConfig containsObject:key]) {
            RCTLogError(@"%@", [NSString stringWithFormat:@"MuxConfig key %@ is required", key]);
            return nil;
        } else {
            return fallback;
        }
    }
}

- (void)Logger:(NSString *)str {
    NSLog(@"==== VIMAI: %@", str);
}

- (void)createCustomerPlayerData:(NSDictionary *)muxConfig {
    customerPlayerData = [[MUXSDKCustomerPlayerData alloc] initWithEnvironmentKey:[self getKeyFrom:muxConfig key:CONFIG_ENVIRONMENT_KEY fallback:nil]];
    NSString* playerName = [self getKeyFrom:muxConfig key:CONFIG_PLAYER_NAME fallback:@"iOS Vimai Player"];
    if (playerName) {
        customerPlayerData.playerName = playerName;
    }
    NSString* userId = [self getKeyFrom:muxConfig key:CONFIG_VIEW_USER_ID fallback:nil];
    if (userId) {
        customerPlayerData.viewerUserId = userId;
    }
    NSString* subPropertyId = [self getKeyFrom:muxConfig key:CONFIG_VIEW_SUBSCRIPTION_DETAIL fallback:nil];
    if (subPropertyId) {
        customerPlayerData.subPropertyId = subPropertyId;
    }
}

- (void)createCustomerVideoData:(NSDictionary *)muxConfig {
    customerVideoData = [MUXSDKCustomerVideoData new];
    NSString* videoTitle = [self getKeyFrom:muxConfig key:CONFIG_VIDEO_TITLE fallback:nil];
    if (videoTitle) {
        customerVideoData.videoTitle = videoTitle;
    }
    NSString* videoContentType = [self getKeyFrom:muxConfig key:CONFIG_VIDEO_CONTENT_TYPE fallback:nil];
    if (videoContentType) {
        customerVideoData.videoContentType = videoContentType;
    }
    NSString* videoLanguageCode = [self getKeyFrom:muxConfig key:CONFIG_VIDEO_LANGUAGE_CODE fallback:nil];
    if (videoLanguageCode) {
        customerVideoData.videoLanguageCode = videoLanguageCode;
    }
}

- (void)createCustomerViewData:(NSDictionary *)muxConfig {
    customerViewData = [MUXSDKCustomerViewData new];
    NSString* viewSessionId = [self getKeyFrom:muxConfig key:CONFIG_VIDEO_LANGUAGE_CODE fallback:nil];
    if (viewSessionId) {
        customerViewData.viewSessionId = viewSessionId;
    }
}

- (instancetype)initWithConfig:(NSDictionary *)muxConfig {
    [self createCustomerPlayerData:muxConfig];
    [self createCustomerVideoData:muxConfig];
    [self createCustomerViewData:muxConfig];
    return self;
}

- (void)startTracking:(AVPlayerLayer *)playerLayer {
    if (!customerPlayerData || !customerVideoData) {
        RCTLogError(@"MuxTracking need initialize. Disable tracking");
        return;
    }
    [self Logger:[NSString stringWithFormat:@"Mux start tracking player %@ on video %@", customerPlayerData.playerName, customerVideoData.videoTitle]];
    [MUXSDKStats monitorAVPlayerLayer:playerLayer
                       withPlayerName:customerPlayerData.playerName
                           playerData:customerPlayerData
                            videoData:customerVideoData
                             viewData:customerViewData];
}

@end
